### DISCUSSION TITLE
bad optional access error

### DISCUSSION THREAD

**Discussion by junhyeok-yoon** [ANLAB-KAIST/KENSv3#142] [Project Q&A 2. Transport Layer (TCP)] [answered]
When I call sendPacket("IPv4", std::move(packet)), I am encoutering 
"unknown file: Failure
C++ exception with description "bad optional access" thrown in the test body."
this error. I do not know where this error occur. But I think that the packet is not properly constructed.
1. Should I also fill the IP header and Ethernet header also?
(PA#2 document에서 [Figuring out the source IP address]라는 tip에 들어가면 주석에 Inside TCPAssignment class ~ 하면서 routingInfoInterface의 함수들을 사용하는 것으로 보아 필요해 보임)
2. If so, why we need to fill those information which is not a role for TCP layer?
3. I want to know where I should use NIC port and mac.
Thanks.

**Comment by Pusnow** [comment, marked as answer]
> 1. Should I also fill the IP header and Ethernet header also?

You don't need to fill all of IP/Ethernet fields.
You only need to specify source and destination IP addresses.

> 2. If so, why we need to fill those information which is not a role for TCP layer?

In practical networking stack implementations, there is no clear boundaries across layers.
In KENS, the TCP layer specifies source/destination IP addresses, since IP layer cannot decide IP addresses without help from TCP layer.

> 3. I want to know where I should use NIC port and mac.

To know source IP address, you should use `getRoutingTable` whose return value is NIC port, then you can use it to retrieve source IP address. 

You don't need to use MAC address directly in PAs. It is used internally.
